File: leetcode/medium/BinaryTreeLevelOrderTraversal.py
from typing import Optional, List
import collections

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        Performs a level order traversal of a binary tree (Breadth-First Search).

        This solution uses a queue to keep track of nodes to visit at each level.
        Using `collections.deque` is crucial for an efficient O(1) pop operation
        from the left of the queue.

        Algorithm:
        1. Handle the edge case of an empty tree.
        2. Initialize a queue (`collections.deque`) with the root node.
        3. Initialize an empty list `result` to store the levels.
        4. Loop as long as the queue is not empty:
           a. Get the number of nodes at the current level (`len(queue)`).
           b. Create an empty list `level` to store the values of nodes at this level.
           c. Loop `len(queue)` times to process all nodes for the current level.
              i. Dequeue a node from the front of the queue.
              ii. Append its value to the `level` list.
              iii. If the node has left/right children, enqueue them.
           d. Append the `level` list to the `result`.
        5. Return the `result`.

        Time Complexity: O(N), where N is the number of nodes, as each node is visited once.
        Space Complexity: O(W), where W is the maximum width of the tree. In the worst
        case (a complete binary tree), this can be O(N).
        """
        if not root:
            return []

        result = []
        # Use collections.deque for an efficient O(1) queue implementation.
        queue = collections.deque([root])

        while queue:
            # Get the number of nodes at the current level.
            level_size = len(queue)
            level = []

            # Process all nodes for the current level.
            for _ in range(level_size):
                # Dequeue the node from the front. O(1) operation.
                node = queue.popleft()
                level.append(node.val)

                # Enqueue children for the next level.
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(level)
        return result

File: leetcode/medium/BinaryTreeLevelOrderTraversal.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/**
 * Solution for performing a level-order traversal (Breadth-First Search) on a binary tree.
 */
class Solution {
    /**
     * Returns the values of the nodes in a binary tree, level by level.
     *
     * This implementation uses a queue to perform a standard Breadth-First Search (BFS).
     * It processes one tree level at a time before moving to the next.
     *
     * Time Complexity: O(N), where N is the number of nodes in the tree. Each node is visited exactly once.
     * Space Complexity: O(M), where M is the maximum number of nodes at any single level
     *                  (the maximum width of the tree). In the worst case, this is O(N).
     *
     * @param root The root node of the binary tree.
     * @return A list of lists, where each inner list contains the values of the nodes at that level.
     */
    public List<List<Integer>> levelOrder(TreeNode root) {
        // Optimization 1: Handle the edge case of an empty tree.
        // If the root is null, there are no levels to traverse, so we return an empty list.
        // This prevents a NullPointerException later in the code.
        if (root == null) {
            return Collections.emptyList();
        }

        // The main list to store the result, containing a list for each level.
        List<List<Integer>> result = new ArrayList<>();

        // A queue is the ideal data structure for BFS. We use ArrayDeque for its efficiency.
        Queue<TreeNode> queue = new ArrayDeque<>();

        // Start by adding the root node to the queue.
        queue.offer(root);

        // Continue the loop as long as there are nodes to process in the current level.
        while (!queue.isEmpty()) {
            // Create a new list to store the values of the nodes for the current level.
            List<Integer> currentLevel = new ArrayList<>();

            // Get the number of nodes in the current level. This is crucial because
            // we only want to process these nodes in this iteration, not the children
            // we are about to add to the queue.
            int levelSize = queue.size();

            // Iterate through all nodes of the current level.
            for (int i = 0; i < levelSize; i++) {
                // Remove the node from the front of the queue.
                TreeNode currentNode = queue.poll();

                // Add the node's value to our current level's list.
                currentLevel.add(currentNode.val);

                // Optimization 2: Add children to the queue for the next level's processing.
                // We check for null to avoid adding empty nodes to the queue.
                if (currentNode.left != null) {
                    queue.offer(currentNode.left);
                }
                if (currentNode.right != null) {
                    queue.offer(currentNode.right);
                }
            }

            // After processing all nodes in the current level, add the list of values to our final result.
            result.add(currentLevel);
        }

        return result;
    }
}